src/core/comm/coll.cc

```diff
@@ -0,0 +1,451 @@
+                         nullptr,
+                         global_comm->global_comm_size);
+    thread_comms[global_comm->unique_id].buffers =
+      (const void**)malloc(sizeof(void*) * global_comm_size);
+    thread_comms[global_comm->unique_id].displs =
+      (const int**)malloc(sizeof(int*) * global_comm_size);
+    for (int i = 0; i < global_comm_size; i++) {
+      thread_comms[global_comm->unique_id].buffers[i] = nullptr;
+      thread_comms[global_comm->unique_id].displs[i]  = nullptr;
+    }
+    __sync_synchronize();
+    thread_comms[global_comm->unique_id].ready_flag = true;
+  }
+  __sync_synchronize();
+  volatile ThreadComm* data = &(thread_comms[global_comm->unique_id]);
+  while (data->ready_flag != true) { data = &(thread_comms[global_comm->unique_id]); }
+  global_comm->comm = &(thread_comms[global_comm->unique_id]);
+  collBarrierLocal(global_comm);
+  assert(global_comm->comm->ready_flag == true);
+  assert(global_comm->comm->buffers != nullptr);
+  assert(global_comm->comm->displs != nullptr);
+#endif
+  if (global_comm->global_comm_size % global_comm->mpi_comm_size == 0) {
+    global_comm->nb_threads = global_comm->global_comm_size / global_comm->mpi_comm_size;
+  } else {
+    global_comm->nb_threads = global_comm->global_comm_size / global_comm->mpi_comm_size + 1;
+  }
+  return CollSuccess;
+}
+
+int collCommDestroy(CollComm global_comm)
+{
+#if defined(LEGATE_USE_GASNET)
+  if (global_comm->mapping_table.global_rank != nullptr) {
+    free(global_comm->mapping_table.global_rank);
+    global_comm->mapping_table.global_rank = nullptr;
+  }
+  if (global_comm->mapping_table.mpi_rank != nullptr) {
+    free(global_comm->mapping_table.mpi_rank);
+    global_comm->mapping_table.mpi_rank = nullptr;
+  }
+#else
+  collBarrierLocal(global_comm);
+  if (global_comm->global_rank == 0) {
+    pthread_barrier_destroy((pthread_barrier_t*)&(thread_comms[global_comm->unique_id].barrier));
+    free(thread_comms[global_comm->unique_id].buffers);
+    thread_comms[global_comm->unique_id].buffers = nullptr;
+    free(thread_comms[global_comm->unique_id].displs);
+    thread_comms[global_comm->unique_id].displs = nullptr;
+    __sync_synchronize();
+    thread_comms[global_comm->unique_id].ready_flag = false;
+  }
+  __sync_synchronize();
+  volatile ThreadComm* data = &(thread_comms[global_comm->unique_id]);
+  while (data->ready_flag != false) { data = &(thread_comms[global_comm->unique_id]); }
+#endif
+  global_comm->status = false;
+  return CollSuccess;
+}
+
+int collAlltoallv(const void* sendbuf,
+                  const int sendcounts[],
+                  const int sdispls[],
+                  CollDataType sendtype,
+                  void* recvbuf,
+                  const int recvcounts[],
+                  const int rdispls[],
+                  CollDataType recvtype,
+                  CollComm global_comm)
+{
+  log_coll.print("Alltoallv: global_rank %d, mpi_rank %d, unique_id %d, comm_size %d",
```

Comment:
I don't think `print` is the right logging level for things like this. it should be `debug` instead.

Reply:
I plan to leave it as print for now until we finalize the PR. I will mark it.

Reply:
Let's fix it now and remove the WIP prefix from the title. I think this PR is almost ready for the merge.

Reply:
fixed.